File: 2015/lua/day03.lua
local f = io.open("../inputs/day03.txt", "r")

input = f:read("a")
local count = 1
local x = 1
local y = 1
local visited = {[1] = {[1] = true}}

function have_visited(x, y)
  if not visited[x] then 
    visited[x] = {[y] = true}
    return 1
  end
  if not visited[x][y] then
    visited[x][y] = true
    return 1
  end
  return 0
end

for idx = 1, #input do
  if input:byte(idx) == string.byte("^") then
    x = x + 1
    count = count + have_visited(x, y)
  elseif input:byte(idx) == string.byte("v") then
    x = x - 1
    count = count + have_visited(x, y)
  elseif input:byte(idx) == string.byte(">") then
    y = y + 1
    count = count + have_visited(x, y)
  elseif input:byte(idx) == string.byte("<") then
    y = y - 1
    count = count + have_visited(x, y)
  else
    print("Char not found: ", input:byte(idx))
  end
end
print("Part 1:", count)

 -- part 2
local soln2 = 1
local x2 = 1
local y2 = 1
local visited = {[1] = {[1] = true}}
for idx = 1, #input, 2 do
  if input:byte(idx) == string.byte("^") then
    x = x + 1
    soln2 = soln2 + have_visited(x, y)
  elseif input:byte(idx) == string.byte("v") then
    x = x - 1
    soln2 = soln2 + have_visited(x, y)
  elseif input:byte(idx) == string.byte(">") then
    y = y + 1
    soln2 = soln2 + have_visited(x, y)
  elseif input:byte(idx) == string.byte("<") then
    y = y - 1
    soln2 = soln2 + have_visited(x, y)
  else
    print("Char not found: ", input:byte(idx))
  end

  if input:byte(idx+1) == string.byte("^") then
    x2 = x2 + 1
    soln2 = soln2 + have_visited(x2, y2)
  elseif input:byte(idx+1) == string.byte("v") then
    x2 = x2 - 1
    soln2 = soln2 + have_visited(x2, y2)
  elseif input:byte(idx+1) == string.byte(">") then
    y2 = y2 + 1
    soln2 = soln2 + have_visited(x2, y2)
  elseif input:byte(idx+1) == string.byte("<") then
    y2 = y2 - 1
    soln2 = soln2 + have_visited(x2, y2)
  else
    print("Char not found: ", input:byte(idx+1))
  end
end

print("Part 2:", soln2)
f:close()

File: 2022/fortran/app/day03.f90
! Advent of Code Solutions
! Copyright (C) 2022  Isaac Curtis
program day03
  use iso_fortran_env
  use aoc_2022_fortran
  implicit none

  integer(int32) :: iunit,ierr,fid,nread
  character(:), allocatable    ::  line, first, second
  character(len=1024) :: buffer, ioerrmsg
  integer(int64) :: part1, part2, n_chars, i, elves, seen1, seen2, group_elves(3)
  call initialize_day
  
  part1 = 0
  part2 = 0
  elves = 1

  nread = 0
  open(newunit=fid,file="../inputs/day03.txt", status='old')
  do
    ! Very long custom read section, because fotran kind of sucks at i/o
    buffer = ''
    line = ''
    do
      read(fid,'(A)',advance='NO',size=nread,iostat=ierr,iomsg=ioerrmsg) buffer
      if (is_iostat_eor(ierr)) then
        ! add the last block of text before the end of record
        if (nread>0) line = line//buffer(1:nread)
        if (nread.eq.0) goto 20 ! if line is empty, stop reading
        exit
      else if (is_iostat_end(ierr)) then
        goto 20
      else if (ierr==0) then ! all the characters were read
        line = line//buffer  ! add this block of text to the string
      else  ! some kind of error
        write(*,*) ierr, ioerrmsg
        error stop 'Read error'
        exit
      end if
    end do

    seen1 = 0
    seen2 = 0
    n_chars = len(line)/2
    first = line(1:n_chars); second = line(n_chars+1:)
    do i = 1, n_chars
      seen1 = ibset(seen1, value_branchless(first(i:i)))
      seen2 = ibset(seen2, value_branchless(second(i:i)))
    end do
    part1 = part1 + trailz(iand(seen1, seen2))
    group_elves(elves) = ior(seen1, seen2)
    if (elves == 3) then
      part2 = part2 + trailz(iand(group_elves(1), iand(group_elves(2), group_elves(3))))
    end if
    elves = merge(1_int64, elves+1, elves==3)
  end do
  20 close(fid)

  write(*,*) "Running Day03"
  write(*,"(a,i10)") "  part 1: ", part1
  write(*,"(a,i10)") "  part 2: ", part2
  call time_day

  contains
  integer function value(c1)
    character(len=1), intent(in) :: c1
    if(ichar(c1) >= ichar('a')) then
      value = 1 + ichar(c1) - ichar('a')
    else
      value = 26 + 1 + ichar(c1) - ichar('A')
    end if
  end function value

  integer function value_branchless(c1)
    character(len=1), intent(in) :: c1
    integer :: cint, c
    value_branchless = mod(ichar(c1) - 38, 58)
  end function value_branchless

end program
